R/setup_computation.R

```diff
@@ -632,6 +633,57 @@ create_S_batch_new <- function(internal, seed = NULL) {
+    # in said lines.
+    # We can now use the additional (new) parameter
+    # `n_unique_approaches = internal$parameters$n_unique_approaches`.
+    # So instead of doing
+    # `pmax(1, round(.N / (n_combinations - 2) * n_batches))`
+    # one could maybe do something like
+    # `round(.N / (n_combinations - 2) * (n_batches - n_unique_approaches)) + 1`.
+    # Here we subtract `n_unique_approaches` as we know that at least `n_unique_approaches` of the
+    # `n_batches` have been looked to a specific approach, so we only want to divide the remaining
+    # batches among the approaches. We add 1 as each method needs to have 1 batch, and these
+    # corresponds to the `n_unique_approaches` batches we subtracted before.
+    # But this is will break too.
+    # Consider same example as in `demonstrate_combined_appraoches_bugs.R`.
+    # There `n_combinations = 32`, `n_unique_approaches = 2`, and `.N = c(5, 25)`.
+    # If we let `n_batches = 5`, then my proposal breaks, as
+    # round(.N / (n_combinations - 2) * (n_batches - n_unique_approaches)) + 1
+    # gives c(1,3) which sums to 4 < 5.
+    # This is because before we round we have c(0.5, 2.5) which are both rounded down
+    # So my conclusion is that it might be the easiest to do what is done above,
+    # or use my proposed approach and add batches until the correct amount has been reached.
+    # Discuss with Martin.
+    # Furthermore, we can both end up in situations with too few and too many coalitions,
+    # so have to check for both.
+    # Consider the same situation where one has `n_batches = 15`, and `n_combinations = 32`, then
+    # round(c(5, 25) / (n_combinations - 2) * n_batches)
+    # will yield c(2,12), whose sum is larger less `n_batches`
+
+    # Ensures that the number of batches corresponds to `n_batches`
+    if (sum(batch_count_dt$n_batches_per_approach) != n_batches) {
+      # Ensure that the number of batches is not larger than `n_batches`.
+      # Remove one batch from the approach with the most batches.
+      while (sum(batch_count_dt$n_batches_per_approach) > n_batches) {
+        approach_to_subtract_batch <- which.max(batch_count_dt$n_batches_per_approach)
+        batch_count_dt$n_batches_per_approach[approach_to_subtract_batch] <-
+          batch_count_dt$n_batches_per_approach[approach_to_subtract_batch] - 1
+      }
+
+      # Ensure that the number of batches is not lower than `n_batches`.
+      # Add one batch to the approach with most coalitions per batch
+      while (sum(batch_count_dt$n_batches_per_approach) < n_batches) {
+        approach_to_add_batch <- which.max(batch_count_dt$n_S_per_approach /
+                                            batch_count_dt$n_batches_per_approach)
+        batch_count_dt$n_batches_per_approach[approach_to_add_batch] <-
+          batch_count_dt$n_batches_per_approach[approach_to_add_batch] + 1
+      }
+    }
+
```

Comment:
@LHBO Not sure whether this extra complexity is really worth it to cover the edge cases where the n_batches does not add up. An alternative is to provide a warning that the number of batches is changed (if specified byt he user), or otherwise just change it. What do you think?

Reply:
@martinju, I agree that it is not very elegant code, but it does not take long to run and is only run once.
The two main reasons I added this is:

1. The first reason was to ensure that the number of batches adds up to `n_batches`.
2. The second reason, and likely the most important, is that if the user has specified `future::plan(multisession, workers = n_batches)`, then it would be nice that `shapr`actually used `n_batches`.

If this is not that important, then I am okay with just giving a warning.